klippa/src/serialize.rs

```diff
@@ -188,6 +207,106 @@ impl Serializer {
         Ok(())
     }
 
+    pub(crate) fn pop_pack(&mut self, share: bool) -> Option<ObjIdx> {
+        self.current?;
+
+        // Allow cleanup when we've error'd out on int overflows which don't compromise the serializer state
+        if self.in_error() && !self.only_overflow() {
+            return None;
+        }
+
+        let pool_idx = self.current.unwrap();
+        let len;
+        let obj_head;
+        {
+            // code logic is a bit different from Harfbuzz serializer here
+            // because I need to move code around to fix mutable borrow/immutable borrow issue
+            let obj = self.object_pool.get_obj_mut(pool_idx)?;
+
+            self.current = obj.next_obj;
+            obj.tail = self.head;
+            obj.next_obj = None;
+
+            if obj.tail <= obj.head {
+                self.set_err(SerializeErrorFlags::SERIALIZE_ERROR_OTHER);
+                return None;
+            }
+
+            len = obj.tail - obj.head;
+
+            // TODO: consider zerocopy
+            // Rewind head
+            self.head = obj.head;
+            self.tail -= len;
+
+            obj_head = obj.head;
+            obj.head = self.tail;
+            obj.tail = self.tail + len;
+
+            //TODO: consider zerocopy here
+            self.data.copy_within(obj_head..obj_head + len, self.tail);
+        }
+
+        let mut hash = 0_u64;
+        // introduce flags to avoid mutable borrow and immutable borrow issues
+        let mut obj_duplicate = false;
+        let mut dup_obj_idxes = (0, 0);
+        if share {
+            hash = hash_one_pool_idx(pool_idx, &self.data, &self.object_pool);
+            if let Some(entry) =
+                self.packed_map
+                    .get_with_hash(pool_idx, hash, &self.data, &self.object_pool)
+            {
+                obj_duplicate = true;
+                dup_obj_idxes = *entry;
+            };
+        }
+
+        if obj_duplicate {
+            self.merge_virtual_links(pool_idx, dup_obj_idxes);
+            let obj = self.object_pool.get_obj_mut(pool_idx).unwrap();
+            obj.fini();
+            self.object_pool.release(pool_idx);
```

Comment:
could we have release automatically call .fini()?

Reply:
good idea! We should do the same thing in harfbuzz as well.